[Merged by Bors] - feat(Geometry/Euclidean/Sphere/Basic): add theorems for angles subtended by diameters

Scarlett-le opened the pull request:
Add two formulations relating sphere diameters and right angles:

* `angle_eq_pi_div_two_iff_mem_sphere_of_isDiameter`: For a diameter of a sphere, 
  the angle subtended by the diameter at any other point on the sphere is a right 
  angle if and only if the point lies on the sphere.

* `angle_eq_pi_div_two_iff_mem_sphere_ofDiameter`: For three distinct points, 
  the angle at the second point is a right angle if and only if the second point 
  lies on the sphere having the first and third points as diameter endpoints.

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Scarlett-le committed:
feat(Geometry/Euclidean/Sphere/Basic): add theorems for angles subtended by diameters

Scarlett-le committed:
fix:add missing import

Scarlett-le committed:
fix:format code

jsm28 commented on Mathlib/Geometry/Euclidean/Sphere/Basic.lean:
```diff
@@ -7,6 +7,7 @@ import Mathlib.Analysis.Convex.StrictConvexBetween
 import Mathlib.Analysis.InnerProductSpace.Convex
 import Mathlib.Analysis.Normed.Affine.Convex
 import Mathlib.Geometry.Euclidean.Basic
+import Mathlib.Geometry.Euclidean.Angle.Unoriented.Affine
```
Please put lemmas about angles and spheres in `Mathlib.Geometry.Euclidean.Angle.Sphere` rather than increasing imports in this file.

I encourage adding a statement that can be explicitly described as Thales's theorem (and then adding it to the 1000-theorems list, `docs/1000.yaml`, as such).

jsm28 replied:
Well, the lemmas should go in a section of `Mathlib.Geometry.Euclidean.Angle.Sphere` that doesn't have `variable`s that assume an orientation or a two-dimensional space (creating a new section there as needed). The only versions that might be appropriate in `Sphere.Basic` are ones expressed purely in terms of inner products and not mentioning angles at all (from which versions with angles could then be deduced in `Angle.Sphere`).

Scarlett-le replied:
Got it! I'll move the theorems to `Mathlib.Geometry.Euclidean.Angle.Sphere` and create a new section without the 2D/orientation constraints. Thanks for the clarification!

Scarlett-le committed:
feat(Geometry/Euclidean/Angle/Sphere): add Thales' theorem and related resultss

Scarlett-le committed:
feat(docs): add Thales's theorem to 1000.yaml

Scarlett-le commented:
@jsm28  I've addressed your feedback by moving the theorems to `Mathlib.Geometry.Euclidean.Angle.Sphere` and adding the entry to `docs/1000.yaml`. The CI checks are now passing. Please let me know if there are any other changes needed.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,81 @@ end Orientation
+is a right angle if and only if the second point lies on the sphere having the first and third
+points as diameter endpoints. -/
+theorem angle_eq_pi_div_two_iff_mem_sphere_ofDiameter (p₁ p₂ p₃ : P) :
+    ∠ p₁ p₂ p₃ = π / 2 ↔ p₂ ∈ Sphere.ofDiameter p₁ p₃ := by
+  exact angle_eq_pi_div_two_iff_mem_sphere_of_isDiameter p₁ p₂ p₃
```
It's unusual to need `by exact`; normally you should just remove both `by` and `exact` and use the term directly.

Scarlett-le replied:
Fixed, thanks!

Scarlett-le committed:
Remove `_ℝ` subscripts and fix formatting in sphere angle theorems

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,81 @@ end Orientation
+
+/-- For a diameter of a sphere, the angle subtended by the diameter
+at any other point on the sphere is a right angle. -/
+theorem angle_eq_pi_div_two_iff_mem_sphere_of_isDiameter (p₁ p₂ p₃ : P) (s : Sphere P)
+    (hd : s.IsDiameter p₁ p₃) :
+    ∠ p₁ p₂ p₃ = π / 2 ↔ p₂ ∈ s := by
```
It's usual for `iff` lemmas that hypotheses used on both side of the `iff` (so `p₂` here) are implicit `{}` rather than explicit `()`. It's also usual that if one hypothesis can be deduced from the type of another, it's implicit; that would mean `p₁`, `p₃` and `s` are implicit because they can be deduced from `hd`. (All of this unless there is a concrete reason to think that explicit arguments make the lemma more usable in practice.)

Scarlett-le replied:
Thanks for all the feedback on the explicit parameter conventions! I'll update the theorems accordingly , making parameters implicit where they follow the usual `iff` lemma conventions. I appreciate the guidance as I'm still learning these mathlib style conventions.

jsm28 commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,81 @@ end Orientation
+/-- For three distinct points, the angle at the second point
+is a right angle if and only if the second point lies on the sphere having the first and third
+points as diameter endpoints. -/
+theorem angle_eq_pi_div_two_iff_mem_sphere_ofDiameter (p₁ p₂ p₃ : P) :
+    ∠ p₁ p₂ p₃ = π / 2 ↔ p₂ ∈ Sphere.ofDiameter p₁ p₃ :=
```
Similarly, all the hypotheses would be implicit based on usual conventions for `iff` lemmas.

Scarlett-le committed:
refactor: change explicit parameters to implicit in angle theorems

Scarlett-le commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,81 @@ end Orientation
+
+/-- **Thales' theorem**: The angle inscribed in a semicircle is a right angle. -/
+theorem thales_theorem (p₁ p₂ p₃ : P) (s : Sphere P) (hd : s.IsDiameter p₁ p₃) :
+    ∠ p₁ p₂ p₃ = π / 2 ↔ p₂ ∈ s :=
```
Thanks, too.

JovanGerb commented:
I had a go at golfing the main proof. Here is my shorter proof:

```
theorem angle_eq_pi_div_two_iff_mem_sphere_of_isDiameter {p₁ p₂ p₃ : P} {s : Sphere P}
    (hd : s.IsDiameter p₁ p₃) :
    ∠ p₁ p₂ p₃ = π / 2 ↔ p₂ ∈ s := by
  rw [mem_sphere', EuclideanGeometry.angle,
    ← InnerProductGeometry.inner_eq_zero_iff_angle_eq_pi_div_two]
  let o := s.center
  have h_center : o = midpoint ℝ p₁ p₃ := hd.midpoint_eq_center.symm
  rw [← vsub_add_vsub_cancel p₁ o p₂, ← vsub_add_vsub_cancel p₃ o p₂,
    inner_add_left, inner_add_right, inner_add_right]
  have h_opp : p₁ -ᵥ o = -(p₃ -ᵥ o) := by
    rw [h_center, left_vsub_midpoint, right_vsub_midpoint, ← smul_neg, neg_vsub_eq_vsub_rev]
  rw [h_opp, inner_neg_left, inner_neg_left, real_inner_comm (p₃ -ᵥ o) (o -ᵥ p₂)]
  ring_nf
  rw [neg_add_eq_zero, real_inner_self_eq_norm_sq, ← dist_eq_norm_vsub,
    real_inner_self_eq_norm_sq, ← dist_eq_norm_vsub, sq_eq_sq₀ dist_nonneg dist_nonneg,
    mem_sphere.mp hd.right_mem]
  exact eq_comm
```

JovanGerb commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,80 @@ end Orientation
+  angle_eq_pi_div_two_iff_mem_sphere_of_isDiameter (Sphere.isDiameter_ofDiameter p₁ p₃)
+
+/-- **Thales' theorem**: The angle inscribed in a semicircle is a right angle. -/
+theorem thales_theorem {p₁ p₂ p₃ : P} {s : Sphere P} (hd : s.IsDiameter p₁ p₃) :
```
Is this exactly the same as the above theorem? So can it be defined using `alias`?

jcommelin commented on Mathlib/Geometry/Euclidean/Angle/Sphere.lean:
```diff
@@ -66,6 +66,80 @@ end Orientation
+
+open Real InnerProductSpace
+
+/-- For a diameter of a sphere, the angle subtended by the diameter
+at any other point on the sphere is a right angle. -/
```
Please mention Thales here.

Scarlett-le committed:
refactor: simplify Thales' theorem proof and improve documentation

jcommelin reviewed (commented):
LGTM. One more comment about a docstring. Otherwise, let's get this merged.

bors d+

JovanGerb commented:
(Note also my comment above about the redundant closing and opening of the same namespace)

Scarlett-le committed:
update
